src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs

```diff
@@ -325,6 +346,53 @@ public Task<IAsyncCursor<RawBsonDocument>> ResumeAsync(IReadBinding binding, Can
             return aggregateOperation.ExecuteAsync(binding, cancellationToken);
         }
 
+        /// <inheritdoc />
+        public void ReviewResumeParams()
+        {
+            // Docs: if the ChangeStream has a cached postBatchResumeToken and has returned all documents in the most recent batch(or the most recent batch was empty):
+            // The driver MUST set resumeAfter to the cached postBatchResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo: should be implemented in scope of the CSHARP-2514.
+
+            // Docs: if the ChangeStream has a cached documentResumeToken:
+            // The driver MUST set resumeAfter to the cached documentResumeToken.
+            // The driver MUST NOT set startAfter. If startAfter was in the original aggregation command, the driver MUST remove it.
+            // The driver MUST NOT set startAtOperationTime. If startAtOperationTime was in the original aggregation command, the driver MUST remove it.
+            // todo:
+
+            // Docs: if the ChangeStream was created with startAfter:
+            // The driver MUST set resumeAfter to the value of the originally used startAfter.
+            // The driver MUST NOT set startAfter. The driver MUST remove it from the original aggregation command.
+            // The driver MUST NOT set startAtOperationTime.
+            if (_startAfter != null)
+            {
+                _resumeAfter = _startAfter;
+                _startAfter = null;
+                _startAtOperationTime = null;
```

Comment:
Let's not mutate the state of the class. That has the potential to make debugging harder because it becomes harder to tell whether the state was set by the caller or mutated by this method.

It would be fine to have a method that throws exceptions if incompatible options are set.

And we can have methods that compute (without mutating state) the "effective" values to use when building commands to send to the server.

Reply:
As I understand, it's a requirement from the spec. 
>It would be fine to have a method that throws exceptions if incompatible options are set.

In spec, there is a requirement that we should not throw an exception in case if  incompatible options are set.

>And we can have methods that compute (without mutating state) the "effective" values to use when building commands to send to the server.

Do you mean to have a method which will take `ChangeStreamOperation` as an income parameter and return the modified object?

Reply:
No. I just don't want to see the fields of the object mutated.

We can compute the effective resumeAfter or startAfter (etc..) instead of mutating the fields of the class.